File: MediaLibrary.Web/src/app/color.service.ts
import { Injectable } from '@angular/core';

export type RGB = readonly [number, number, number];
export type RGBA = readonly [number, number, number, number];

type Colorish = string | RGB | RGBA;
function isRGB(value: Colorish): value is (RGB | RGBA) {
    return Array.isArray(value);
}

@Injectable({
    providedIn: 'root'
})
export class ColorService {
    public readonly black = '#000';
    public readonly white = '#fff';
    public static readonly GAMMA = 1.8;

    private ctx: CanvasRenderingContext2D;
    private cache: { [color: string]: RGBA } = {};

    constructor() {
        const canvas = document.createElement('canvas');
        canvas.width = canvas.height = 1;
        this.ctx = canvas.getContext('2d');
    }

    public contrast(a: Colorish, b: Colorish): number {
        return this.contrastRatio(
            this.contrastBrightness(a),
            this.contrastBrightness(b));
    }

    public contrastBrightness(color: Colorish, gamma?: number): number {
        if (!isRGB(color)) {
            color = this.parseColor(color);
        }

        if (typeof gamma === "undefined") {
            gamma = ColorService.GAMMA
        }

        return 0.25 * Math.pow(color[0] / 255, gamma * 2) +
            0.54 * Math.pow(color[1] / 255, gamma * 2) +
            0.21 * Math.pow(color[2] / 255, gamma * 2);
    }

    public contrastColor(color: Colorish): string {
        if (!color) {
            return this.black;
        }

        if (!isRGB(color)) {
            this.parseColor(color);
        }

        const brightness = this.contrastBrightness(color);
        const whiteContrast = this.contrastRatio(brightness, this.contrastBrightness([255, 255, 255]));
        const blackContrast = this.contrastRatio(brightness, this.contrastBrightness([0, 0, 0]));
        return whiteContrast > blackContrast ? this.white : this.black;
    }

    public parseColor(color: string): RGBA {
        if (Object.prototype.hasOwnProperty.call(this.cache, color)) {
            return this.cache[color];
        }

        // From: https://stackoverflow.com/a/19366389
        this.ctx.clearRect(0, 0, 1, 1);
        this.ctx.fillStyle = this.black;
        this.ctx.fillStyle = color;
        const computed = this.ctx.fillStyle;
        this.ctx.fillStyle = this.white;
        this.ctx.fillStyle = color;
        if (computed === this.ctx.fillStyle) {
            this.ctx.fillRect(0, 0, 1, 1);
            var d = this.ctx.getImageData(0, 0, 1, 1).data;
            return this.cache[color] = [d[0], d[1], d[2], d[3]];
        } else {
            return this.cache[color] = undefined;
        }
    }

    private contrastRatio(l1: number, l2: number): number {
        return (Math.max(l1, l2) + 0.25) / (Math.min(l1, l2) + 0.25);
    }
}

File: MediaLibrary/Services/ColorService.cs
namespace MediaLibrary.Services
{
    using System;
    using System.Collections.Generic;
    using System.Drawing;
    using System.Linq;

    public static class ColorService
    {
        public const double Gamma = 1.8;

        private static Dictionary<string, Color?> cache = new Dictionary<string, Color?>();

        public static Color Blend(double amount, Color a, Color b)
        {
            return Blend((amount, a), (1 - amount, b));
        }

        public static Color Blend(IEnumerable<Color> colors)
        {
            return Blend(colors.Select(c => (1D, c)));
        }

        public static Color Blend(params Color[] colors)
        {
            return Blend(colors.Select(c => (1D, c)));
        }

        public static Color Blend(params (double, Color)[] colors)
        {
            return Blend(colors.AsEnumerable());
        }

        public static Color Blend(IEnumerable<(double weight, Color value)> colors)
        {
            var r = 0.0;
            var g = 0.0;
            var b = 0.0;
            var weight = 0.0;

            foreach (var color in colors)
            {
                r += Math.Pow(color.value.R, Gamma) * color.weight;
                g += Math.Pow(color.value.G, Gamma) * color.weight;
                b += Math.Pow(color.value.B, Gamma) * color.weight;
                weight += color.weight;
            }

            return Color.FromArgb(
                (int)Math.Round(Math.Pow(r / weight, 1 / Gamma)),
                (int)Math.Round(Math.Pow(g / weight, 1 / Gamma)),
                (int)Math.Round(Math.Pow(b / weight, 1 / Gamma)));
        }

        public static double Contrast(Color a, Color b) =>
            ContrastRatio(
                ContrastBrightness(a),
                ContrastBrightness(b));

        public static double ContrastBrightness(Color color) =>
            ContrastBrightness(color, Gamma);

        public static double ContrastBrightness(Color color, double gamma) =>
            0.25 * Math.Pow(color.R / 255.0, gamma * 2) +
            0.54 * Math.Pow(color.G / 255.0, gamma * 2) +
            0.21 * Math.Pow(color.B / 255.0, gamma * 2);

        public static Color ContrastColor(Color? color) =>
            color is Color c ? ContrastColor(c) : Color.Black;

        public static Color ContrastColor(Color color)
        {
            var brightness = ContrastBrightness(color);
            var whiteContrast = ContrastRatio(brightness, ContrastBrightness(Color.White));
            var blackContrast = ContrastRatio(brightness, ContrastBrightness(Color.Black));
            return whiteContrast > blackContrast ? Color.White : Color.Black;
        }

        public static Color? ParseColor(string color)
        {
            if (string.IsNullOrEmpty(color))
            {
                return null;
            }

            lock (cache)
            {
                if (cache.TryGetValue(color, out var cached))
                {
                    return cached;
                }
            }

            Color? value;
            try
            {
                value = new ColorParser().Parse(color.Trim());
            }
            catch (FormatException)
            {
                value = null;
            }

            lock (cache)
            {
                return cache[color] = value;
            }
        }

        private static double ContrastRatio(double l1, double l2)
        {
            return (Math.Max(l1, l2) + 0.25) / (Math.Min(l1, l2) + 0.25);
        }
    }
}
